substrate/runtime/contract/src/vm.rs

```diff
@@ -0,0 +1,785 @@
+			.with_forbidden_floats();
+
+		let module = self
+			.module
+			.take()
+			.expect("On entry to the function `module` can't be `None`; qed");
+
+		let contract_module = pwasm_utils::inject_gas_counter(module, &gas_rules)
+			.map_err(|_| Error::GasInstrumentation)?;
+
+		self.module = Some(contract_module);
+		Ok(())
+	}
+
+	fn inject_stack_height_metering(&mut self) -> Result<(), Error> {
+		let module = self
+			.module
+			.take()
+			.expect("On entry to the function `module` can't be `None`; qed");
+
+		let contract_module =
+			pwasm_utils::stack_height::inject_limiter(module, self.config.max_stack_height)
+				.map_err(|_| Error::StackHeightInstrumentation)?;
+
+		self.module = Some(contract_module);
+		Ok(())
+	}
+
+	/// Find the memory import entry and return it's descriptor.
+	fn find_mem_import(&self) -> Option<&MemoryType> {
+		let import_section = self
+			.module
+			.as_ref()
+			.expect("On entry to the function `module` can't be `None`; qed")
+			.import_section()?;
+		for import in import_section.entries() {
+			if let ("env", "memory", &External::Memory(ref memory_type)) =
+				(import.module(), import.field(), import.external())
+			{
+				return Some(memory_type);
+			}
+		}
+		None
+	}
+
+	fn into_wasm_code(mut self) -> Result<Vec<u8>, Error> {
+		elements::serialize(
+			self.module
+				.take()
+				.expect("On entry to the function `module` can't be `None`; qed"),
+		).map_err(|_| Error::Serialization)
+	}
+}
+
+struct PreparedContract {
+	instrumented_code: Vec<u8>,
+	memory: sandbox::Memory,
+}
+
+fn prepare_contract<T: Trait>(original_code: &[u8], config: &Config<T>) -> Result<PreparedContract, Error> {
+	let mut contract_module = ContractModule::new(original_code, config)?;
+	contract_module.ensure_no_internal_memory()?;
+	contract_module.inject_gas_metering()?;
+	contract_module.inject_stack_height_metering()?;
+
+	// Inspect the module to extract the initial and maximum page count.
+	let memory = match contract_module.find_mem_import() {
+		Some(memory_type) => {
+			let limits = memory_type.limits();
+			match (limits.initial(), limits.maximum()) {
+				(initial, Some(maximum)) if initial > maximum => {
+					// Requested initial number of pages should not exceed the requested maximum.
+					return Err(Error::Memory);
+				}
+				(_, Some(maximum)) if maximum > config.max_memory_pages => {
+					// Maximum number of pages should not exceed the configured maximum.
+					return Err(Error::Memory);
+				}
+				(_, None) => {
+					// Maximum number of pages should be always declared.
+					// This isn't a hard requirement and can be treated as a maxiumum set
+					// to configured maximum.
+					return Err(Error::Memory);
+				}
+				(initial, maximum) => sandbox::Memory::new(initial, maximum),
+			}
+		}
+
+		// If none memory imported then just crate an empty placeholder.
+		// Any access to it will lead to out of bounds trap.
+		None => sandbox::Memory::new(0, Some(0)),
+	}.map_err(|_| Error::Memory)?;
```

Comment:
I don't understand how this could happen.

Reply:
that comes from the fact that `sandbox::Memory::new` could return an error which has a wasmi specific type. So we convert it into a value of the local  `Error` and then `?` it.

Reply:
Cleaned that, please check out!